src/mlpack/methods/ann/layer/concat_layer.hpp

```diff
@@ -0,0 +1,183 @@
+/**
+ * @file concat_layer.hpp
+ * @author Nilay Jain
+ * @author Marcus Edel
+ *
+ * Definition of the ConcatLayer class.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_CONCAT_LAYER_HPP
+#define MLPACK_METHODS_ANN_LAYER_CONCAT_LAYER_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/activation_functions/rectifier_function.hpp>
+#include <mlpack/methods/ann/layer/layer_traits.hpp>
+#include <mlpack/methods/ann/layer/one_hot_layer.hpp>
+#include <mlpack/methods/ann/layer/conv_layer.hpp>
+#include <mlpack/methods/ann/layer/pooling_layer.hpp>
+#include <mlpack/methods/ann/layer/softmax_layer.hpp>
+#include <mlpack/methods/ann/layer/bias_layer.hpp>
+#include <mlpack/methods/ann/layer/linear_layer.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+/**
+ * @tparam JoinLayers Contains all layer modules that need be concatenated.
+ */
+template <typename JoinLayers,
+          typename InputDataType = arma::cube,
+          typename OutputDataType = arma::cube>
+class ConcatLayer
+{
+ public:
+  ConcatLayer(const size_t numLayers,
+              JoinLayers &&layers) :
+  numLayers(numLayers),
+  layers(std::forward<JoinLayers>(layers))
+  {
+    /* Nothing to do. */
+  }
+
+  template<typename eT>
+  void Forward(const arma::Cube<eT>& , arma::Cube<eT>& output)
+  {
+    ForwardTail(layers, output);
+  }
+
+  template<size_t I = 0, typename DataType, typename... Tp>
+  typename std::enable_if<I == sizeof...(Tp), void>::type
+  ForwardTail(std::tuple<Tp...>& layers, DataType& output)
+  {
+    /* Nothing to do. */
+  }
+
+  template<size_t I = 0, typename DataType, typename... Tp>
+  typename std::enable_if<I < sizeof...(Tp), void>::type
+  ForwardTail(std::tuple<Tp...>& layers, DataType& output)
+  {
+    output = arma::join_slices(output, std::get<I>(layers).OutputParameter());
```

Comment:
The problem I see here is, that it only works if we use cubes as output type. We could use template specializations to distinguish between arma::mat and arma::cube what do you think?